<a href="https://colab.research.google.com/github/nicolemartins7/Prova-Integrada_temp/blob/master/IA_N2_2024S1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<head>
  <meta name="author" content="Rogério de Oliveira">
  <meta institution="author" content="Universidade Presbiteriana Mackenzie">
</head>

<img src="http://meusite.mackenzie.br/rogerio/mackenzie_logo/UPM.2_horizontal_vermelho.jpg" width=300, align="right">
<!-- <h1 align=left><font size = 6, style="color:rgb(200,0,0)"> optional title </font></h1> -->


In [103]:
# Nicole Martins dos Santos

In [ ]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
%matplotlib inline
import seaborn as sns

# Case: **Student Stress**

Neste case você vai empregar dados relacionados a estudantes para determinar o seu nível de stress.

In [ ]:
df_train = pd.read_csv('https://github.com/Rogerio-mack/IA_2024S1/raw/main/data/StressStudentsTrain.csv')
display(df_train.head(), df_train.shape)

df_test = pd.read_csv('https://github.com/Rogerio-mack/IA_2024S1/raw/main/data/StressStudentsTest.csv')
display(df_test.head(), df_test.shape)

anxiety_level  self_esteem  mental_health_history  depression  headache  \
0              0           28                      0           3         1   
1             14           30                      0           9         2   
2             21           11                      1          20         3   
3             21            4                      1          26         3   
4              3            0                      1           4         1   

   blood_pressure  sleep_quality  breathing_problem  noise_level  \
0               2              4                  2            1   
1               3              5                  5            1   
2               3              1                  4            5   
3               3              1                  3            3   
4               3              1                  5            4   

   living_conditions  ...  basic_needs  academic_performance  study_load  \
0                  3  ...            5                     4           2   
1                  4  ...            3                     5           0   
2                  1  ...            1                     1           3   
3                  2  ...            2                     2           5   
4                  3  ...            3                     1           1   

   teacher_student_relationship  future_career_concerns  social_support  \
0                             5                       1               3   
1                             0                       4               1   
2                             2                       5               1   
3                             1                       5               1   
4                             2                       1               1   

   peer_pressure  extracurricular_activities  bullying  stress_level  
0              1                           1         1             0  
1              2                           3         0             0  
2              5                           5         4             2  
3              5                           4         4             2  
4              0                           4         4             2  

[5 rows x 21 columns]

(880, 21)

anxiety_level  self_esteem  mental_health_history  depression  headache  \
0              7           30                      0           3         1   
1             15            0                      1          27         3   
2             14           19                      1          10         2   
3             16            4                      1          26         4   
4              7           29                      0           5         1   

   blood_pressure  sleep_quality  breathing_problem  noise_level  \
0               2              5                  2            1   
1               3              1                  5            5   
2               1              2                  4            3   
3               3              1                  4            5   
4               2              4                  1            1   

   living_conditions  ...  basic_needs  academic_performance  study_load  \
0                  4  ...            5                     5           1   
1                  2  ...            1                     1           4   
2                  2  ...            3                     2           2   
3                  2  ...            1                     2           5   
4                  3  ...            4                     5           1   

   teacher_student_relationship  future_career_concerns  social_support  \
0                             4                       1               3   
1                             2                       4               1   
2                             2                       2               2   
3                             1                       4               1   
4                             5                       1               3   

   peer_pressure  extracurricular_activities  bullying  stress_level  
0              2                           1         1             0  
1              4                           4         4             2  
2              2                           3         3             1  
3              5                           5         4             2  
4              2                           1         1             0  

[5 rows x 21 columns]

(220, 21)

# Exercício 1. **Decision Tree Classification $\times$ Multilayer Perceptron**

Neste exercício você fará um modelo de árvore de decisão e outro de MLP para predição o nível de stress dos estudantes.

Para normalização dos dados empregue o `StandardScaler`. Em todos lugares onde for necessário para reprodutibilidade dos resultados, empregue o `random_state=1`.





## Exercício 1.1.

Separe os atributos preditores e o atributo objetivo dos conjuntos de treinamento e teste e, em seguida, normalize os dados.

**Dica:** normalize os dados de treinamento e aplique o mesmo estimador ajustado aos dados de teste!

In [86]:
from sklearn.preprocessing import StandardScaler
scaler =  StandardScaler()


X_train = df_train.drop(columns=['stress_level'])
y_train = df_train['stress_level']

X_test = df_test.drop(columns=['stress_level'])
y_test = df_test['stress_level']



scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)


In [87]:
#@markdown check if TRUE
X_train_scaled.sum().sum() == -2.957634137601417e-13 and X_test_scaled.sum().sum() == -3.0458031228583113

True

## Exercício 1.2.

Faça um modelo de Árvore de Decisão. Exceto o `random_state`, empregue todos os demais parâmetros padrão.

**Q1. Qual a acuracidade do seu modelo de Árvore de Decisão obtida sobre o conjunto de teste?**

In [88]:
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import accuracy_score

dt_model = DecisionTreeClassifier(random_state=1)
dt_model.fit(X_train_scaled, y_train)

y_pred = dt_model.predict(X_test_scaled)

accuracy = accuracy_score(y_test, y_pred)
print("Acurácia Árvore:", accuracy)


Acurácia Árvore: 0.8909090909090909


## Exercício 1.3.

**Q2. Use a árvore de decisão obtida para descobrir: quais os 3 atributos apresentam maior ganho de informação para determinar o nível de stress dos estudantes?**

In [101]:
feature_importance = dt_model.feature_importances_

importance_df = pd.DataFrame({'Feature': X_train.columns, 'Importance': feature_importance})
importance_df = importance_df.sort_values(by='Importance', ascending=False)

top_3_features = importance_df.head(3)['Feature'].tolist()
print( top_3_features)


['blood_pressure', 'self_esteem', 'bullying']


## Exercício 1.4.

Faça um modelo Multi-Layer Perceptron. Empregue uma rede neural com 3 camadas ocultas com 4,8,4 elementos respectivamente. Configure ainda a tangente hiperbólica como função de ativação e 1000 iterações de treinamento do modelo.

**Q3. Qual a acuracidade do seu modelo MLP obtida sobre o conjunto de teste?**

In [90]:
from sklearn.neural_network import MLPClassifier
from sklearn.metrics import accuracy_score

mlp_model = MLPClassifier(hidden_layer_sizes=(4, 8, 4), activation='tanh', max_iter=1000, random_state=1)
mlp_model.fit(X_train_scaled, y_train)

y_pred_mlp = mlp_model.predict(X_test_scaled)

accuracy_mlp = accuracy_score(y_test, y_pred_mlp)
print("Acurácia  MLP:", accuracy_mlp)

Acurácia  MLP: 0.8954545454545455


## Exercício 1.5.

**Q4. No modelo de maior acuracidade, qual classe apresenta melhores resultados, e qual o seu f1-score?**

**Q5. Qual o número de falsos negativos e falsos positivos para a classe da questão anterior?**

In [91]:
accuracy # acucaria arvore

0.8909090909090909

In [92]:
accuracy_mlp # acucaria MLP

0.8954545454545455

In [93]:
conf_matrix = confusion_matrix(y_test, y_pred_mlp)
conf_matrix

array([[66,  4,  4],
       [ 2, 65,  5],
       [ 7,  1, 66]])

In [94]:
class_report = classification_report(y_test, y_pred_mlp)
print(class_report)

              precision    recall  f1-score   support

           0       0.88      0.89      0.89        74
           1       0.93      0.90      0.92        72
           2       0.88      0.89      0.89        74

    accuracy                           0.90       220
   macro avg       0.90      0.90      0.90       220
weighted avg       0.90      0.90      0.90       220



In [95]:
fn, fp = conf_matrix[1][0], conf_matrix[0][1]

In [102]:
print("maior acurácia: MLP", accuracy_mlp)
print("F1 classe 1:  0.92 " )
print("N falsos negativos:", fn)
print("N  falsos positivos:", fp)

maior acurácia: MLP 0.8954545454545455
F1 classe 1:  0.92 
N falsos negativos: 2
N  falsos positivos: 4


# Exercício 2. **Clustering $\times$ Test**

Neste exercício você vai aplicar algoritmos de clusterização Kmédias e Hierárquica sobre o conjunto de teste.

Para normalização dos dados empregue o `StandardScaler`, caso você ainda não tenha feito, faça o exercício 1.1. Em todos lugares onde for necessário para reprodutibilidade dos resultados, empregue o `random_state=1`.







## Exercício 2.1.

Faça uma clusterização kmédias em 3 grupos do conjunto de teste (sem a classe `stress_level`!) e analise os grupos formados. (empregue também o parâmetro `n_init=10`).

**Q6. Qual dos grupos parece apresentar indivíduos com uma saúde geral muito melhor (menos dor de cabeça, maior autoestima, menos depressão, maior qualidade do sono, etc.) e qual grupo apresenta indicadores menores no geral (seja para os indicadores positivos ou negativos)?**

In [97]:
from sklearn.cluster import KMeans

kmeans = KMeans(n_clusters=3, n_init=10, random_state=1)
kmeans.fit(X_test_scaled)

labels = kmeans.labels_

df_test_clustered = df_test.copy()
df_test_clustered['Cluster'] = labels

cluster_summary = df_test_clustered.groupby('Cluster').mean()
cluster_summary


anxiety_level  self_esteem  mental_health_history  depression  \
Cluster                                                                  
0             4.015873    27.428571                0.00000    4.174603   
1            11.228261    17.945652                0.51087   12.500000   
2            17.938462     7.923077                1.00000   21.338462   

         headache  blood_pressure  sleep_quality  breathing_problem  \
Cluster                                                               
0        1.000000        2.015873       4.603175           1.476190   
1        2.543478        1.630435       2.543478           2.782609   
2        3.907692        3.000000       1.015385           3.753846   

         noise_level  living_conditions  ...  basic_needs  \
Cluster                                  ...                
0           1.444444           3.619048  ...     4.428571   
1           2.576087           2.467391  ...     2.304348   
2           4.030769           1.569231  ...     1.446154   

         academic_performance  study_load  teacher_student_relationship  \
Cluster                                                                   
0                    4.412698    1.460317                      4.476190   
1                    2.576087    2.467391                      2.239130   
2                    1.507692    3.984615                      1.384615   

         future_career_concerns  social_support  peer_pressure  \
Cluster                                                          
0                      1.015873        2.968254       1.603175   
1                      2.434783        1.923913       2.423913   
2                      4.461538        0.984615       4.415385   

         extracurricular_activities  bullying  stress_level  
Cluster                                                      
0                          1.555556  1.031746      0.000000  
1                          2.402174  2.347826      1.000000  
2                          4.384615  4.323077      1.969231  

[3 rows x 21 columns]

In [98]:
print("melhor: grupo 0")
print("menores: grupo 0")

melhor: grupo 0
menores: grupo 0


## Exercício 2.2.

Faça agora uma clusterização hierárquica empregando o linkage `complete` e outra empregando o linkage padrão `ward`. Ambas criando 3 grupos cada.



**Q7. A que grupos pertencem o 10 e 15 elementos do conjunto de teste no cluster hierárquico complete e ward respectivamente?**